lib/train/transports/ssh.rb

```diff
@@ -86,16 +87,30 @@ def connection(state = {}, &block)
       end
     end
 
-    # Returns the ssh config option like user, port from config files
-    # Params options [Hash], option_type [String]
-    # Return String
-    def self.read_options_from_ssh_config(options, option_type)
-      config_options = Net::SSH.configuration_for(options[:host], true)
-      config_options[option_type]
+    def apply_ssh_config(host)
+      if options[:ssh_config] != false && !options[:ssh_config].nil?
+        files = options[:ssh_config] == true ? Net::SSH::Config.default_files : options[:ssh_config]
+        host_cfg = ssh_config_for_host(host, files)
+        host_cfg.each do |key, value|
+          if key == :keys && options[:key_files].nil? && !host_cfg[:keys].nil? && options[:password].nil?
+            options[:key_files] = host_cfg[key]
+          else
+            # Give precedence to config file when ssh_config options is set to true or to the path of the config file.
+            # This is required as there are default values set for some of the opitons and we unable to
+            # identify whether the values are set from the cli option or those are default so either we should give
+            # precedence to config file or otherwise we need to check each options default values and then set the value for that option.
```

Comment:
I understand why this is, but this is likely to cause some pain and confusion. We'll need to document this carefully in the consuming CLI applications.